alot/commands/__init__.py

```diff
@@ -10,20 +10,53 @@
 import re
 
 from ..settings import settings
-from ..helper import split_commandstring, string_decode
+from ..helper import split_commandstring, string_decode, classproperty
 
 
 class Command(object):
 
     """base class for commands"""
     repeatable = False
+    mode = ""
+    cmdname = ""
+
+    @classproperty
+    def id(cls):
+        """Mode and command name of instance as string."""
+        return "{}.{}".format(cls.mode, cls.cmdname)
 
     def __init__(self):
-        self.prehook = None
-        self.posthook = None
+        # set to False if no attempt to fetch hooks from settings. After
+        # looking for hooks, they will contain a callable or None if no
+        # hook was found.
+        self._prehook = False
+        self._posthook = False
+
         self.undoable = False
         self.help = self.__doc__
 
+    @property
+    def prehook(self):
+        """Command pre hook."""
+        if self._prehook is False:
+            get_hook = settings.get_hook
+            mode = type(self).mode
+            cmdname = type(self).cmdname
+            self._prehook = (get_hook('pre_%s_%s' % (mode, cmdname)) or
+                             get_hook('pre_global_%s' % cmdname))
+        return self._prehook
```

Comment:
@dcbaker I tried that first, but the function is referenced from a class attribute (I had them `_prehook` and `_posthook` on the same level as `mode` and `cmdname`).

When the hooks were called, Python treated them as if the hook was a function defined on that level, i.e. expected them to receive `self` as their first argument which is not how the hook functions are defined.

Here is my working solution:

- Hook class: https://github.com/fnurl/alot/blob/hook-always-class-store/alot/settings/manager.py#L467
- hooks as `classproperties`: https://github.com/fnurl/alot/blob/hook-always-class-store/alot/commands/__init__.py#L26
- wrapping the hooks: https://github.com/fnurl/alot/blob/hook-always-class-store/alot/settings/manager.py#L311